### DISCUSSION TITLE
Can't generate views into another directory

### DISCUSSION THREAD

**Discussion by holdenhinkle** [janko/rodauth-rails#51] [Q&A] [answered]
Hi @janko ,

I'm setting up `rodauth-rails` for the first time. I'm thrilled with how easy it's been so far. Thank you for creating this Gem!

I'm writing because I'm not able to generate the `rodauth` views in another directory.

When I run `rails generate rodauth:views --name admin`, the following views are generated:

```
√ dating-app % rails generate rodauth:views --name admin
The dependency tzinfo-data (>= 0) will be unused by any of the platforms Bundler is installing for. Bundler is installing for ruby but the dependency is only for x86-mingw32, x86-mswin32, x64-mingw32, java. To add those platforms to the bundle, run `bundle lock --add-platform x86-mingw32 x86-mswin32 x64-mingw32 java`.
Running via Spring preloader in process 205
      create  app/views/rodauth/_field.html.erb
      create  app/views/rodauth/_field_error.html.erb
      create  app/views/rodauth/_login_field.html.erb
      create  app/views/rodauth/_login_display.html.erb
      create  app/views/rodauth/_password_field.html.erb
      create  app/views/rodauth/_submit.html.erb
      create  app/views/rodauth/_login_form.html.erb
      create  app/views/rodauth/_login_form_footer.html.erb
      create  app/views/rodauth/_login_form_header.html.erb
      create  app/views/rodauth/login.html.erb
      create  app/views/rodauth/multi_phase_login.html.erb
      create  app/views/rodauth/logout.html.erb
      create  app/views/rodauth/_login_confirm_field.html.erb
      create  app/views/rodauth/_password_confirm_field.html.erb
      create  app/views/rodauth/create_account.html.erb
      create  app/views/rodauth/_login_hidden_field.html.erb
      create  app/views/rodauth/verify_account_resend.html.erb
      create  app/views/rodauth/verify_account.html.erb
      create  app/views/rodauth/reset_password_request.html.erb
      create  app/views/rodauth/reset_password.html.erb
      create  app/views/rodauth/_new_password_field.html.erb
      create  app/views/rodauth/change_password.html.erb
      create  app/views/rodauth/change_login.html.erb
      create  app/views/rodauth/close_account.html.erb
```

I expected the views to be generated into `app/views/admin`.

Before running that generator I set the the controller for the `admin` configuration to:
```ruby
    rails_controller { Admin::RodauthController }
```

Here's the controller which is in `/app/controllers/admin/rodauth_controller.rb:
```ruby
module Admin
  class RodauthController < ApplicationController
    # used by Rodauth for rendering views, CSRF protection, and running any
    # registered action callbacks and rescue_from handlers
    layout :rodauth_layout

    private
  
    def rodauth_layout
      case request.path
      when rodauth(:admin).login_path,
           rodauth(:admin).create_account_path,
           rodauth(:admin).verify_account_path,
           rodauth(:admin).reset_password_path,
           rodauth(:admin).reset_password_request_path
        "authentication"
      else
        "admin"
      end
    end
  end
end
```

I copied the generated views from `/views/rodauth` over to the `/views/admin` but they're not being picked up when I hit `http://localhost:3000/admin/login` -- the builtin `rodauth` templates are being rendered.

Can you please tell me what I might be doing wrong here? Or what I'm missing?

Thank you,

Holden

**Comment by janko** [comment, marked as answer]
The documentation was showing an incorrect option name for `rodauth:views`, it should be `--directory` instead of `--name`.

The reason the templates weren't picked up is that `Admin::RodauthController` will look for templates inside `app/views/admin/rodauth`, not in `app/views/admin`.

**Reply by holdenhinkle** [reply]
Ah, perfect. Thank you for your fast response, Janko!

**Reply by holdenhinkle** [reply]
These links are now pointing to the root directory:

```
Create a New Account
Forgot Password?
Resend Verify Account Information
```

They were pointing to `http://localhost:3000/admin/` before.

Do I need to update them in this `each` block in `app/views/admin/rodauth/_login_form_footer.html.erb`?

```ruby
    <% rodauth.login_form_footer_links.sort.each do |_, link, text| %>
      <li><%= link_to text, link %></li>
    <% end %>
```

Or should they be working like before? (Meaning, something is wrong.)

Update...

I think something is wrong with the paths that are being generated. When I tried submitting the login form I got this error:

```
Routing Error
No route matches [POST] "/login"
Rails.root: /Users/holdenhinkle/Documents/Projects/dating-app
```

**Reply by holdenhinkle** [reply]
Btw, when I remove the `/views/admin/rodauth/` directory, the paths that are generated in the builtin templates are correct...

**Reply by janko** [reply]
Aha, in admin templates you'll need to replace all references of `rodauth`, which uses the main configuration, with `rodauth(:admin)`, which uses the admin configuration.

I'll also improve the generator so that this is done automatically.

**Reply by holdenhinkle** [reply]
Awesome, thank you Janko! It's working perfectly now.

**Reply by holdenhinkle** [reply]
I'm totally in love with `rodauth` and `rodauth-rails`. Thank you Jeremy and Janko!!
